### PR TITLE
multiple policy issues

### PR DESCRIPTION
**Pull request by Connoropolous**
this should actually prbly be hotfixed into metamaps.cc as well

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -65,6 +65,10 @@ def as_json_for_autocomplete
     json
   end
   
+  def all_accessible_maps
+    maps + shared_maps
```
Comment (devvmh) [inline app/models/user.rb]: `maps.or(shared_maps)` is better because it stays as an activerecord relation

File: app/policies/mapping_policy.rb [hunk 2]
```diff
@@ -8,11 +8,11 @@ def resolve
...
-        scope.joins(:map).where(permission, visible).or(scope.joins(:map).where(user_id: user.id))
-      else
-        scope.joins(:map).where(permission, visible)
-      end
+      return scope.joins(:map).where(permission, visible) unless user
+
+      scope.joins(:map).where(permission, visible)
+           .or(scope.joins(:map).where('maps.id IN (?)', user.shared_maps.map(&:id)))
+           .or(scope.joins(:map).where('maps.user_id = ?', user.id))
     end
```
Comment (Connoropolous) [inline app/policies/mapping_policy.rb]: So the policy_scope for `mappings` and `messages` is the same... it is entirely based off the map to which it belongs

File: app/policies/synapse_policy.rb [hunk 3]
```diff
@@ -7,7 +7,7 @@ def resolve
       return scope.where(permission: visible) unless user
 
       scope.where(permission: visible)
-           .or(scope.where(defer_to_map_id: user.shared_maps.map(&:id)))
+           .or(scope.where(defer_to_map_id: user.all_accessible_maps.map(&:id)))
```
Comment (devvmh) [inline app/policies/synapse_policy.rb]: brilliant. can we add a comment explaining this mind bending permission logic?

File: app/policies/synapse_policy.rb [hunk 4]
```diff
@@ -7,7 +7,7 @@ def resolve
       return scope.where(permission: visible) unless user
```
Comment (Connoropolous) [inline app/policies/synapse_policy.rb]: in the topics and synapses policy_scopes, these changes are to fix the circumstances where:
you want to see a topic or synapse which
- is not `commons`
- you did not create
- has a `defer_to_map_id` value which references a map that is not "shared with you" but that you created

File: app/policies/synapse_policy.rb (hunk 3 above)
Comment (devvmh) [inline app/policies/synapse_policy.rb]: maybe something like "map is visible if 1) it's public/commons, 2) it's defer_to_map_id isn't nil and points to a map visible to the user, or 3) the user owns the map"

maybe we should even make that note about not being nil explicit in the code

    .or(scope.where.not(defer_to_map_id: nil).where(defer_to_map_id: user.all_accessible_maps.map(&:id))

File: app/models/user.rb (hunk 1 above)
Comment (Connoropolous) [inline app/models/user.rb]: this somehow combines the two? even though its an 'or'?

**Pull request merged**